symengine/solve.cpp

```diff
@@ -0,0 +1,304 @@
+        // substitute x = y-a/4 to get equation of the form y**4 + e*y**2 + f*y
+        // + g = 0
+        auto sqa = mul(a, a);
+        auto cba = mul(sqa, a);
+        auto aby4 = div(a, i4);
+        auto e = sub(b, div(mul(i3, sqa), i8));
+        auto ff = sub(add(c, div(cba, i8)), div(mul(a, b), i2));
+        auto g = sub(add(d, div(mul(sqa, b), i16)),
+                     add(div(mul(a, c), i4), div(mul({i3, cba, a}), i256)));
+
+        // two special cases
+        if (eq(*g, *zero)) {
+            vec_basic newcoeffs(3);
+            newcoeffs[0] = ff, newcoeffs[1] = e, newcoeffs[2] = zero;
+            auto rcubic = solve_poly_cubic(newcoeffs, sym, domain);
+            SYMENGINE_ASSERT(is_a<FiniteSet>(*rcubic));
+            for (auto &r :
+                 down_cast<const FiniteSet &>(*rcubic).get_container()) {
+                roots.insert(sub(r, aby4));
+            }
+            roots.insert(aby4);
+        } else if (eq(*ff, *zero)) {
+            vec_basic newcoeffs(2);
+            newcoeffs[0] = g, newcoeffs[1] = e;
+            auto rquad = solve_poly_quadratic(newcoeffs, sym, domain);
+            SYMENGINE_ASSERT(is_a<FiniteSet>(*rquad));
+            auto rtemp = down_cast<const FiniteSet &>(*rquad).get_container();
+            for (auto &r : rtemp) {
+                auto sqrtr = sqrt(r);
+                roots.insert(sub(sqrtr, aby4));
+                roots.insert(sub(neg(sqrtr), aby4));
+            }
+        } else {
+            // Leonhard Euler's method
+            vec_basic newcoeffs(3);
+            newcoeffs[0] = neg(div(mul(ff, ff), i64)),
+            newcoeffs[1] = div(sub(mul(e, e), mul(i4, g)), i16),
+            newcoeffs[2] = div(e, i2);
+
+            auto rcubic = solve_poly_cubic(newcoeffs, sym);
+            SYMENGINE_ASSERT(is_a<FiniteSet>(*rcubic));
+            roots = down_cast<const FiniteSet &>(*rcubic).get_container();
+            auto p = sqrt(*roots.begin());
+            auto q = p;
+            if (roots.size() != 1) {
+                q = sqrt(*std::next(roots.begin()));
+            }
+            auto r = div(neg(ff), mul({i8, p, q}));
+            roots.clear();
+            roots.insert(add({p, q, r, neg(aby4)}));
+            roots.insert(add({p, neg(q), neg(r), neg(aby4)}));
+            roots.insert(add({neg(p), q, neg(r), neg(aby4)}));
+            roots.insert(add({neg(p), neg(q), r, neg(aby4)}));
+        }
+    }
+    return set_intersection({domain, finiteset(roots)});
+}
+
+RCP<const Set> solve_poly_heuristics(const RCP<const UExprPoly> &f,
+                                     const RCP<const Symbol> &sym,
+                                     const RCP<const Set> &domain)
+{
+    auto degree = f->get_poly().degree();
+    auto lc = f->get_poly().find_cf(degree).get_basic();
+    vec_basic coeffs;
+
+    if (degree > 0 and degree <= 4) {
+        for (int i = 0; i < degree; i++)
+            coeffs.push_back(div(f->get_poly().find_cf(i).get_basic(), lc));
```

Comment:
It's easier to implement this way.
Should I generalise this by doing the ^above within each lower degree poly solver?